### PR TITLE
Add app initialisation options to gfdatasource

### PR DESCRIPTION
**Pull request by leth**
- [x] Update README
- [x] Fix `__`
- [x] Fix sorting of args
- [x] Remove disable flag
- [x] Check subparsers

### CODE REVIEW

**Comment by jml** [comment]
This is kind of hard to review because of the rename. What's the purpose of this change? How will people take advantage of it? (I guess README will answer that). Are the command-line args of current version compatible with this one?

**Comment by leth** [comment]
> This is kind of hard to review because of the rename.

I'll try pulling that out to a separate commit, if not will restore the name.

> What's the purpose of this change?

Grafana has 'plugins' and 'apps'; this allows you to enable an 'app', which in turn enables bundled plugins, which will allow us to add a new data source.

> How will people take advantage of it? (I guess README will answer that). 

Yeah will get to that!

> Are the command-line args of current version compatible with this one?

Should be, yes

File: gfdatasource/gfdatasource [hunk 1]
```diff
@@ -103,19 +235,30 @@ def _split_creds(url):
...
+        cmd_func = _cmds[opts.cmd]
+    except KeyError:
+        print('Unknown command', opts.cmd)
+        sys.exit(1)
+
+    update_func = cmd_func(grafana_api, opts)
+
     while True:
-        grafana_api.update_datasource(datasource)
+        update_func()
```
Comment (jml) [inline gfdatasource/gfdatasource]: Why set `update_func` only to call it here?

File: gfdatasource/gfdatasource [hunk 2]
```diff
@@ -59,33 +88,136 @@ class GrafanaAPI(object):
...
-        help="How frequently to update Grafana, in seconds",
+
+    app_parser = subparsers.add_parser('app')
+    app_parser.add_argument(
+        '--id', type=str, required=True, help="The app plugin ID",
+    )
+    app_parser.add_argument(
+        '--disabled', type=bool, default=False,
+        help="Whether the app should be disabled",
+    )
```
Comment (jml) [inline gfdatasource/gfdatasource]: Why would you want to use this?

File: gfdatasource/gfdatasource [hunk 3]
```diff
@@ -59,33 +88,136 @@ class GrafanaAPI(object):
...
+    __default_subparser = None
+
+    def set_default_subparser(self, name):
+        self.__default_subparser = name
+
+    def parse_args(self, args=None, namespace=None):
+        subparser_action = None
+        for action in self._subparsers._actions:
+            if isinstance(action, argparse._SubParsersAction):
+                subparser_action = action
```
Comment (jml) [inline gfdatasource/gfdatasource]: Why do we want to set this to the last `_SubParsersAction` in `_actions`?

File: gfdatasource/gfdatasource [hunk 4]
```diff
@@ -59,33 +88,136 @@ class GrafanaAPI(object):
...
+    )
+
+    def update_app():
+        grafana_api.update_app(app)
+
+    return update_app
+
+
+class DefaultSubcommandArgParse(argparse.ArgumentParser):
+    __default_subparser = None
```
Comment (jml) [inline gfdatasource/gfdatasource]: Personally, I avoid `__foo` variables, preferring `_foo` instead, as the name-mangling only helps in tiny codebases. However, feel free to leave as is.

File: gfdatasource/gfdatasource [hunk 5]
```diff
@@ -59,33 +88,136 @@ class GrafanaAPI(object):
...
+        secure_json_data=secure_json_data,
+    )
+
+    def update_app():
+        grafana_api.update_app(app)
+
+    return update_app
+
+
+class DefaultSubcommandArgParse(argparse.ArgumentParser):
```
Comment (jml) [inline gfdatasource/gfdatasource]: I don't really understand what this is for.

File: gfdatasource/gfdatasource (hunk 4 above)
Comment (leth) [inline gfdatasource/gfdatasource]: Good spot; I copied the original version from the interwebs.

File: gfdatasource/gfdatasource (hunk 2 above)
Comment (leth) [inline gfdatasource/gfdatasource]: I have no idea! I'll remove it.

File: gfdatasource/gfdatasource (hunk 5 above)
Comment (leth) [inline gfdatasource/gfdatasource]: Previously the command parser understood `<cmd> <*options>`
Now it understands `<cmd> <global opts> [datasource|app] <specific opts>`
This is here to make `<cmd> <global opts> <datasource opts>` default to the original behaviour.

I have now spotted that it'll do a poor job of separating an interleaving of global and datasource options.

File: gfdatasource/gfdatasource (hunk 3 above)
Comment (leth) [inline gfdatasource/gfdatasource]: I think there is currently always only one of them. I'll check this

File: gfdatasource/gfdatasource (hunk 1 above)
Comment (leth) [inline gfdatasource/gfdatasource]: I didn't want to copy the while + sleep loop into both functions, but still avoid duplicate work.

It's probably overcomplicating things.

Comment (jml) [inline gfdatasource/gfdatasource]: Maybe just call `cmd_func(grafana_api, opts)` directly, without the temp var?

**Comment by jml** [comment]
LMK when you want a follow-up review.

**Comment by leth** [comment]
@jml I think it's ready for another look :)

**Pull request merged**